lib/wallaroo/core/common/state_step_creator.pony

```diff
@@ -0,0 +1,286 @@
+            state_name.cstring())
+          Fail()
+          error
+        end
+
+        let id = _step_id_gen()
+        let state_step = Step(_auth, runner_builder(
+          where event_log = _event_log, auth = _auth),
+          consume reporter, id, runner_builder.route_builder(),
+          _event_log, recovery_replayer, _outgoing_boundaries,
+          this
+          where omni_router = omni_router)
+
+        _pending_steps(state_step) = (state_name, key, id)
+        state_step.initializer_initialized(this)
+      end
+    end
+
+  be set_router_registry(router_registry: RouterRegistry) =>
+    _router_registry = router_registry
+
+  be initialize_routes_and_builders(initializer: LocalTopologyInitializer,
+    keys_to_steps: KeyToStepInfoTag[Step] iso,
+    keys_to_step_ids: KeyToStepInfo[StepId] iso,
+    recovery_replayer: RecoveryReplayer,
+    outgoing_boundaries: Map[String, OutgoingBoundary] val,
+    runner_builders: Map[String, RunnerBuilder] val)
+  =>
+    _keys_to_steps = consume keys_to_steps
+    _keys_to_step_ids = consume keys_to_step_ids
+
+    _recovery_replayer = recovery_replayer
+    _outgoing_boundaries = outgoing_boundaries
+
+    _runner_builders = runner_builders
+
+    initializer.report_initialized(this)
+
+  be report_ready_to_work(initializable: Initializable) =>
+    """
+    Handles the `report_ready_to_work` message from a step once it is ready
+    to work.
+    """
+
+    match initializable
+    | let step: Step =>
+      try
+        ifdef "trace" then
+          @printf[I32](("Got a message that a step is ready!\n").cstring())
+        end
+
+        let router_registry = try
+          _router_registry as RouterRegistry
+        else
+          @printf[I32]("StateStepCreator must have a RouterRegistry.\n"
+            .cstring())
+          Fail()
+          return
+        end
+
+        (let state_name, let key, let id) =
+          _pending_steps(step)?
+
+        _keys_to_steps.add(state_name, key, step)
+        _keys_to_step_ids.add(state_name, key, id)
+
+        router_registry.register_state_step(step, state_name, key, id)
+
+      else
+        @printf[I32](("StateStepCreator received report_ready_to_work from " +
+          "an unknown step.\n").cstring())
+        Fail()
+      end
+    else
+      @printf[I32](("StateStepCreator received report_ready_to_work from a " +
+        "non-step initializable.\n").cstring())
+      Fail()
+    end
+
+  // OmniRouter updates
+
+  be add_boundaries(boundaries: Map[String, OutgoingBoundary] val) =>
+    _add_boundaries(boundaries)
+
+  fun ref _add_boundaries(boundaries: Map[String, OutgoingBoundary] val) =>
+    let new_boundaries = recover iso Map[String, OutgoingBoundary] end
+
+    for (worker, boundary) in _outgoing_boundaries.pairs() do
+      new_boundaries(worker) = boundary
+    end
+
+    for (worker, boundary) in boundaries.pairs() do
+      if not new_boundaries.contains(worker) then
```

Comment:
If it's possible the boundaries include one that's updated, we don't want this guard.

Reply:
Removed the boundary.